wiki/Business-Logic.md

```diff
@@ -0,0 +1,378 @@
+# Business Logic
+
+Business Logic is executable code that implements your business rules, by transforming and manipulating application data.
+
+## Purpose
+
+1. Isolate business logic from rendering logic, user-interface logic, and data access logic
+2. Formally define the business operations that users can perform
+3. Allow most user operations to be performed in parallel.
+4. Allow core business logic to be shared between the two frontends
+5. Allow for business logic to be tested independently from other layers of the application
+
+## Methodology
+
+### Parallelism (Scoping and Logical Threading)
+
+At the business level, MODiX is almost entirely a reactionary application. It reacts to two main types of external stimuli: HTTP Requests and Discord Gateway messages. There is also a very small handful of internally-driven stimuli, generated by the use of timers. This is where parallelism comes into play. MODiX processes such "external stimuli" in parallel to other "external stimuli", but within the context of a single "external stimulus" all processing is done in a sequential, albeit asynchronous, manner.
+
+MODiX operates, asynchronously, under the umbrella of the `ThreadPool`, which leverages just a few physical threads, optimized based on available hardware, to execute any arbitrary number of "logical threads" in parallel to each other. A "logical thread" consists of `async` or otherwise `Task`-based method call stacks, chained together by use of `await`. Although they do not execute synchronously in a literal sense, they behave as if they do, in a logical sense. Each asynchronous chunk of a "logical thread" is guaranteed by .NET and the `ThreadPool` to execute in sequence, and can only be pre-empted by other physical or logical threads.
+
+The other mechanism that makes parallelism possible (dare I say, rather easy) is the concept of scoping provided by Microsoft's Dependency Injection model. Scoping allows for some injected dependencies to be instantiated many times, as needed based on context, rather than just once. This is controlled by the creation of `IServiceScope`s through the IoC container (`IServiceProvider`). Each `IServiceScope` gets its own separate `IServiceProvider` that shares some service instances among all other providers (`Singleton` services), but also will create its own instances that exist only within that scope (`Scoped` services). These services then also get disposed when the scope itself is diposed. This is most easily understood by looking at ASP.NET Core, which handles scoping internally, by creating a unique `IServiceScope` for each HTTP request that it receives. Services within that scope are used to process that HTTP request, and when processing is complete, that scope is disposed. Thus, any particular HTTP request can be processed independently of other requests, because it has its own set of service instances to work with, that are independent of the service instances of other requests.
```

Comment:
Typo: "diposed" -> "disposed"

```suggestion
The other mechanism that makes parallelism possible (dare I say, rather easy) is the concept of scoping provided by Microsoft's Dependency Injection model. Scoping allows for some injected dependencies to be instantiated many times, as needed based on context, rather than just once. This is controlled by the creation of `IServiceScope`s through the IoC container (`IServiceProvider`). Each `IServiceScope` gets its own separate `IServiceProvider` that shares some service instances among all other providers (`Singleton` services), but also will create its own instances that exist only within that scope (`Scoped` services). These services then also get disposed when the scope itself is disposed. This is most easily understood by looking at ASP.NET Core, which handles scoping internally, by creating a unique `IServiceScope` for each HTTP request that it receives. Services within that scope are used to process that HTTP request, and when processing is complete, that scope is disposed. Thus, any particular HTTP request can be processed independently of other requests, because it has its own set of service instances to work with, that are independent of the service instances of other requests.
```